src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -1853,6 +1853,11 @@ private static void ComputeDeclarationsInNode(SemanticModel semanticModel, ISymb
             // Mark completion if we successfully executed all actions and only if we are analyzing a span containing the entire syntax node.
             if (success && analysisStateOpt != null && !declarationAnalysisData.IsPartialAnalysis)
             {
+                // Ensure that we do not mark declaration complete/clear state if cancellation was requested.
+                // Other thread(s) might still be executing analysis, and clearing state could lead to corrupt execution
+                // or unknown exceptions.
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
>cancellationToken.ThrowIfCancellationRequested(); [](start = 16, length = 49)

Why is it sufficient to check for cancellation here. What if cancellation happens when we are executing one of the loops below, would clearing state potentially cause similar problem?

Reply:
This is the only place we explicitly clear state that can be accessed by multiple threads, and no code in the loop below (or any calls made in the loop) check for cancellation, so this will address the issue in the Watson. However, I get your point that the correct fix here would be ensure that the below state clearing code executes only after any other threads that are still attempting to do analysis have completed. We do have state tracking logic to identify if any thread is attempting to execute analyzer actions - however changing that is a non-trivial code refactoring and much more risky change. Let me file a tracking issue to address your concern in 16.1.

Reply:
@AlekseyTs Filed https://github.com/dotnet/roslyn/issues/33098 to harden this code further in 16.1. This PR should still address the Watson where we don't even attempt to execute this state clearing code in presence of cancellation.